<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-seq analysis with Bioconductor: Additional material: High throughput RNA-sequencing</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link" style="color: #383838">Beta
            <i aria-hidden="true" class="icon" data-feather="alert-circle" style="color: #001483; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
        </abbr>

      </div>
    </div>
    <div class="selector-container ">


      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/09-Additional-material-HTS.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-seq analysis with Bioconductor
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-seq analysis with Bioconductor
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html"></a></li>
          </ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq analysis with Bioconductor
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 82%" class="percentage">
    82%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 82%" aria-valuenow="82" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/09-Additional-material-HTS.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-intro-to-rnaseq.html">1. Introduction to RNA-seq</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-setup.html">2. RStudio Project and Experimental Data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-import-annotate.html">3. Importing and annotating quantified data into R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-exploratory-qc.html">4. Exploratory analysis and quality control</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-differential-expression.html">5. Differential expression analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-extra-design.html">6. Extra exploration of design matrices</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-gene-set-analysis.html">7. Gene set enrichment analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-next-steps.html">8. Next steps</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        9. Additional material: High throughput RNA-sequencing
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#rna-seq-library-preparation">RNA-seq library preparation</a></li>
<li><a href="#sequencing">Sequencing</a></li>
<li><a href="#fastq-files">fastq files</a></li>
<li><a href="#processing-pipeline">Processing pipeline</a></li>
<li><a href="#alignment">Alignment</a></li>
<li><a href="#counting">Counting</a></li>
<li><a href="#preparing-the-count-matrix">Preparing the count matrix</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-additionnal-material-LM-FDR.html">10. Additional material: Linear models and FDR</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush12">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading12">
        <a href="11-Additional-material-PCA.html">11. Additional material: PCA</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush13">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading13">
        <a href="12-GSEA.html">12. Additional material: GSEA</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li><li><a href="reference.html"></a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="08-next-steps.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="10-additionnal-material-LM-FDR.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="08-next-steps.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Next steps</a>
        <a class="chapter-link float-end" href="10-additionnal-material-LM-FDR.html" rel="next">Next: Additional material:... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Additional material: High throughput RNA-sequencing</h1>
        <p> Last updated on 2025-11-16 |

        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/09-Additional-material-HTS.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li><p>How does high throughput RNA-sequencing work?</p></li>
<li><p>What is a fastq file?</p></li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Explain what RNA-seq is.</li>
<li>Provide an overview of the procedure to go from the raw data to the
read count matrix that will be used for downstream analysis.</li>
</ul></div>
</div>
</div>
</div>
</div>
<blockquote>
<p><em>This episode is based on the course <a href="https://uclouvain-cbio.github.io/WSBIM2122/" class="external-link">Omics data
analysis</a>.</em></p>
</blockquote>
<section><h2 class="section-heading" id="rna-seq-library-preparation">RNA-seq library preparation<a class="anchor" aria-label="anchor" href="#rna-seq-library-preparation"></a></h2>
<hr class="half-width"><p>The starting point of a RNA-seq experiment is the cDNA library
preparation. RNA is first isolated from cells and purified to remove
ribosomal RNA which constitute the majority of total RNA. This can be
done by extracting poly(A) RNA or by depleting rRNA. Purified RNA is
then fragmented, reverse transcribed and adapters are ligated to both
extremities to allow for amplification and sequencing.</p>
<figure><img src="fig/Unstranded_lib_prep.png" width="40%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="section level3">
<h3 id="stranded-versus-unstranded-libraries">Stranded versus unstranded libraries<a class="anchor" aria-label="anchor" href="#stranded-versus-unstranded-libraries"></a></h3>
<p>The library preparation presented above corresponds to an
<strong>unstranded protocol</strong>.</p>
<p>As both cDNA strands will be amplified for sequencing, about half of
the reads will be sequenced from the first cDNA strand, and half of the
reads will be sequenced from the second strand cDNA. But we won’t know
which strand was sequenced! So we lose the information about the
orientation of the transcript.</p>
<p>This can be avoided by using a <strong>stranded protocol</strong>. In
this case, dUTPs are added during the 2nd strand cDNA synthesis and the
newly synthetized cDNA strand is degradated after the ligation step.
This allows to infer the transcript orientation because only the first
cDNA strand will be sequenced. Indeed, if the sequence of the read
corresponds to the the sense strand of DNA, it means that the gene was
originally transcribed in a sense orientation. On the contrary, if the
read is complementary to the sense strand of DNA, it means that it was
originally transcribed in antisense.</p>
<figure><img src="fig/stranded_vs_unstranded.png" width="80%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Strand-specificity leads to lower number of ambiguous reads.</p>
</div>
</section><section><h2 class="section-heading" id="sequencing">Sequencing<a class="anchor" aria-label="anchor" href="#sequencing"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="se-vs-pe-sequencing">SE vs PE sequencing<a class="anchor" aria-label="anchor" href="#se-vs-pe-sequencing"></a></h3>
<p>Once the library is ready, molecules from the library are sequenced
in a high throughput manner to obtain short sequences from one end
(single-end sequencing) or both ends (pair-end sequencing). Single-end
sequencing is cheaper, but Paired-end sequencing improves the ability to
localise the fragment in the genome and resolve mapping close to repeat
regions, resulting in less multimapping reads.</p>
<figure><img src="fig/SE_vs_PE.png" width="70%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
<div class="section level3">
<h3 id="cluster-amplification">cluster amplification<a class="anchor" aria-label="anchor" href="#cluster-amplification"></a></h3>
<p>The sequencing is done by a cluster amplification process.</p>
<p>The library preparation is hybridized to a flow cell coated with
immobilized oligonucleotides that serve as support to hold the DNA
strands in place during sequencing. The templates are copied from the
hybridized primers by 3’ extension using a high-fidelity DNA polymerase.
The original templates are denatured, leaving the copies immobilized on
the flow cell surface.</p>
<p>Immobilized DNA template copies are then amplified by bridge
amplification. The templates loop over to hybridize to adjacent
oligonucleotides. DNA polymerase copies the templates from the
hybridized oligonucleotides, forming dsDNA bridges, which are denatured
to form two ssDNA strands. These two strands loop over and hybridize to
adjacent oligonucleotides and are extended again to form two new dsDNA
loops. The process is repeated on each template by cycles of
denaturation and amplification to create millions of individual, dense
clonal clusters containing ~2,000 molecules.</p>
<p>Each cluster of dsDNA bridges is denatured, and the reverse strand is
removed by specific base cleavage, leaving the forward DNA strand. The
3’-ends of the DNA strands and flow cell-bound oligonucleotides are
blocked to prevent interference with the sequencing reaction. The
sequencing primer is hybridised to the Illumina adapter. Clusters are
ready for sequencing.</p>
<figure><img src="fig/cluster_amplification.png" width="80%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The sequencing process is done by synthesis. A polymerase adds a
fluorescently tagged dNTP to the DNA strand (only one base is able to be
added per round due to the fluorophore acting as a blocking group, but
the blocking group is reversible). Each of the four bases has a unique
fluorophore, and after each round, the machine records which base was
added. The fluorophore is then washed away and the process is
repeated.</p>
<p>For a dynamic view of the cluster amplification process, watch the <a href="https://www.youtube.com/watch?v=HMyCqWhwB8E" class="external-link">Illumina
video</a>.</p>
</div>
</section><section><h2 class="section-heading" id="fastq-files">fastq files<a class="anchor" aria-label="anchor" href="#fastq-files"></a></h2>
<hr class="half-width"><p>The sequence data generated from the sequencer are received in text
files called <code>fastq</code> files. For a single-end run, one fastq
file is created for each sample. For a paired-end run, two separated
fastq files are generated, each containing sequences from one end.</p>
<p>What does a FASTQ file look like?</p>
<figure><img src="fig/fastq_first_lines.png" width="100%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Each entry in a fastq file consists of 4 lines:</p>
<ul><li>A sequence identifier (by convention preceded by ‘@’) with
information about the sequencing run</li>
<li>The sequence</li>
<li>A delimiter, always starting by the sign ‘+’</li>
<li>The base call quality scores. These are Phred scores, encoded using
<a href="https://theasciicode.com.ar" class="external-link">ASCII characters</a> to represent
the numerical quality scores. Each character is assigned a quality score
between 0 and 40. Quality scores represent the probability that the
corresponding nucleotide call is correct.</li>
</ul><figure><img src="fig/phred_score.png" width="80%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<p>What do you think of the quality of the last sequence presented in
the fastq file above?</p>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Explain what RNA-seq is.</li>
<li>Provide an overview of the procedure to go from the raw data to the
read count matrix that will be used for downstream analysis.</li>
</ul></div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="processing-pipeline">Processing pipeline<a class="anchor" aria-label="anchor" href="#processing-pipeline"></a></h2>
<hr class="half-width"><p>The raw reads from fastq files will need to pass through a different
tools to yield ultimately a count table, representing a snapshot of
individual gene expression levels. The execution of this set of tools in
a specified order is commonly referred to as a pipeline.</p>
<figure><img src="fig/pipeline.png" width="30%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="section level3">
<h3 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a></h3>
<p>Of course, no one will analyse the sequences from the fastq file one
by one to check their quality… Rather, a software called <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
can be used. Then, another program such as <a href="http://www.usadellab.org/cms/?page=trimmomatic" class="external-link">Trimmomatics</a>
can help to clean the reads.</p>
<ul><li><strong>FastQC</strong></li>
</ul><p><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
allows to analyse different features of the reads and generates a report
that includes several diagnostic plots to highlight any problem the data
may have.</p>
<p>FastQC is run from the command line:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode r" tabindex="0"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="sc">$</span> fastqc SampleName.fastq</span></code></pre>
</div>
<div id="discussion" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="discussion" class="callout-inner">
<h3 class="callout-title">Discussion:</h3>
<div class="callout-content">
<p>Try to interpret the different plots of this <a href="https://uclouvain-cbio.github.io/WSBIM2122/OD01_10_1_fastqc.html" class="external-link">fastQC
report</a> from a real RNA-seq fastq file.</p>
<p>Another report corresponding to <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/bad_sequence_fastqc.html" class="external-link">bad
quality</a> data can be used as a point of comparison.</p>
</div>
</div>
</div>
<ul><li><strong>Trimmomatics</strong></li>
</ul><p>FastQC gives a global vision the quality of the sample. If it is not
optimal, it is advisable to use <a href="http://www.usadellab.org/cms/?page=trimmomatic" class="external-link">Trimmomatics</a>
software to filter poor quality reads and trim residual adapters.</p>
<p>Trimmomatics has a variety of options to clean the reads. Read the <a href="http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf" class="external-link">manual</a>
for more information. However, a command for Trimmomatics could look
something like the command below. In this case it would perform adapter
removal, and perform a scan of the read with a 10-base wide sliding
window, cutting when the average quality per base drops below 20.</p>
<pre><code>$ java -jar trimmomatic.jar \
  PE \                                  # paired-end
  -threads 4 \                          # number of threads
  SampleName_1.fastq \                  # fastq file for fw reads
  SampleName_2.fastq  \                 # fastq file for rev reads
  SampleName_1_clean.fastq \            # clean fastq for fw reads
  SampleName_1_unpaired.fastq \         # fastq for unpaired remaining fw reads
  SampleName_2_clean.fastq \            # clean fastq for rev reads
  SampleName_2_unpaired.fastq \         # fastq for unpaired remaining rev reads
  ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 \  # cut adapter from the read.
  SLIDINGWINDOW:10:20                   # sliding window trimming</code></pre>
<div id="discussion-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="discussion-1" class="callout-inner">
<h3 class="callout-title">Discussion:</h3>
<div class="callout-content">
<p>How could you check that the trimming performed well?</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="alignment">Alignment<a class="anchor" aria-label="anchor" href="#alignment"></a></h2>
<hr class="half-width"><p>Next step is to map the reads to determine where in the genome the
reads originated from. Many different aligners are available such as <a href="https://ccb.jhu.edu/software/hisat2/manual.shtml" class="external-link">HISAT2</a>, <a href="https://github.com/alexdobin/STAR" class="external-link">STAR</a>, <a href="https://bio-bwa.sourceforge.net" class="external-link">BWA</a>, <a href="https://combine-lab.github.io/salmon/" class="external-link">Salmon</a>, <a href="https://sourceforge.net/projects/subread/" class="external-link">Subread</a>… There is
no gold standard, but some tools are better suited for particular NGS
analyses <span class="citation">[@Baruzzo:2017]</span>… In the case of
RNA-seq, it is important to use an aligner able to handle spliced
reads.</p>
<figure><img src="fig/spliced_alignment.png" width="70%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Here we show an example of what the command could look like when
using <a href="https://daehwankimlab.github.io/hisat2/" class="external-link">HISAT2</a>
aligner, launched with default parameters. In this basic configuration,
mandatory parameters are of course the fastq file(s) and the sequence of
the genome on which the reads have to be aligned. In practical, the
genome’s sequence is not given as it is but in an indexed form <a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. Of course
many of other parameters can be fine tuned (see the <a href="https://daehwankimlab.github.io/hisat2/manual/" class="external-link">manual</a> for
more details). The alignment output file data is a <code>sam</code> file
(see below).</p>
<pre><code>$ hisat2 \
  -p 6 \                        # number of threads
  -x genome_index \             # index for the reference genome
  -1 SampleName_1_clean.fastq \ # clean fastq_file for fw reads
  -2 SampleName_2_clean.fastq \ # clean fastq_file for rev reads
  -S SampleName.sam             # sam file name</code></pre>
<p>It is recommended to pay attention to the HISAT2 report to check the
alignment rate. For a PE-end alignment it could look like this:</p>
<figure><img src="fig/hisat2_report.png" width="50%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p><strong>SAM format</strong></p>
<p>The <a href="https://genome.sph.umich.edu/wiki/SAM" class="external-link">SAM</a> format is
a generic format for storing large nucleotide sequence alignments. In a
SAM file, each entry gives the alignment information for a single read.
Each entry has 11 mandatories fields for essential mapping information
and a variable number of other fields for aligner specific information.
An example of few entries from a SAM file is displayed below.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">head</span> data/example.sam</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>DRR016707.26395247	147	1	19626495	60	101M	=	19626490	-106	GTAACAATGTTATCAGTAATGCTTTAAACTCCAGCACCTGGTTATGCATTTGAAACCAAGTCTGTTTCTTGTTTTGTATTTTCTCTCTGGAAGTTGTAAGG	DDDDDDDDDDDDDEEEEEEEFFFFFFHCHHHJJJHHJJJJJJJJJJJJIHFJJIHJJJJJJJJJJJJJJJJJJJJJJJJJJJJIJJJJHHHHHFFFFFCCC	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:101	YS:i:0	YT:Z:CP	NH:i:1
DRR016707.26395247	99	1	19626490	60	101M	=	19626495	106	CTTCTGTAACAATGTTATCAGTAATGCTTTAAACTCCAGCACCTGGTTATGCATTTGAAACCAAGTCTGTTTCTTGTTTTGTATTTTCTCTCTGGAAGTTG	CCCFFFFFHHHHHJIJJJJIJJJJJJJJJJIJJJJJJJJJJJJJJJHHIHIJJJIJJHIJJJIIIIIIJGIJJJJDHJJJHHHHHHHFFFFFFEDACEEDD	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:101	YS:i:0	YT:Z:CP	NH:i:1
DRR016707.34969377	163	1	19626487	60	101M	=	19626493	107	ATTCTTCTGTAACAATGTTATCAGTAATGCTTTAAACTCCAGCACCTGGTTATGCATTTGAAACCAAGTCTGTTTCTTGTTTTGTATTTTCTCTCTGGAAG	CC@FFFFFHFHFFHGGJCIFJGHIIIIGIIIIJIIJJJIGIIIIIIGII?:BFHGHIIIIGEHHHGIIDGGH@GGIGHJGHHHHCEHDEFDDFFEEECCEC	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:101	YS:i:0	YT:Z:CP	NH:i:1
DRR016707.34969377	83	1	19626493	60	101M	=	19626487	-107	CTGTAACAATGTTATCAGTAATGCTTTAAACTCCAGCACCTGGTTATGCATTTGAAACCAAGTCTGTTTCTTGTTTTGTATTTTCTCTCTGGAAGTTGTAA	CDDDDDDDDDDDDECEEEEEFFFFD?=HHHHGIIGEGIGIIIHIGDIHHGGGIIHGHIIIIGIIIGIGEIIGGGIIIIGIIIIGHDIIFHHHHFFBDD?C@	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:101	YS:i:0	YT:Z:CP	NH:i:1
DRR016707.42561021	161	1	19626495	60	100M1S	=	19626494	0	GTAACAATGTTATCAGTAATGCTTTAAACTAAAGCACCTGGTTATGCATTTGAAACCAAGTCTGTTTCTTGTTTTGTATTTTCTCTCTGGAAGTTGTAAGA	?8?B1:BDD,,2&lt;:C&lt;2A&lt;&lt;,+A?CBFFAE&lt;+&lt;+2A?EEBD99:D&lt;*?B:*:D9B:DD/BDED@)=8=B;@=)).=(=@;7A)7.==&gt;;3);@AAA&gt;A@3;	AS:i:-8	XN:i:0	XM:i:2	XO:i:0	XG:i:0	NM:i:2	MD:Z:30C0C68	YT:Z:UP	NH:i:1
DRR016707.42561021	81	1	19626494	60	101M	=	19626495	0	TGTAACAATGTTATCAGTAATGCTTTAAACTCCAGCACCTGGTTATGCATTTGACACCAAGTCTGTTTCTTGTTTTGTATTTTCGCTCTGGAAGTTGTAAG	(9(??==3??&gt;==;5(55((6.((9&gt;;..;67&gt;)&gt;).5(A;;&gt;=/))8.=&gt;9)((?AB?92=00*;=::1)=:3=74=A7&lt;)0&lt;+&lt;AAAB?=&lt;+4AA&lt;;==	AS:i:-5	ZS:i:-19	XN:i:0	XM:i:2	XO:i:0	XG:i:0	NM:i:2	MD:Z:54A29T16	YT:Z:UP	NH:i:1
DRR016707.43465443	163	1	19626483	60	101M	=	19626486	104	TGCCATTCTTCTGTAACAATGTTATCAGTAATGCTTTAAACTCCAGCACCTGGTTATGCATTTGAAACCAAGTCTGTTTCTTGTTTTGTATTTTCTCTCTG	B@@FAFFFHHHHHHIIIIJJJHJJIJJIIJJJJJJJFJJJJJJJJJJJJJHIJFHJJIJJJIJJIIJIJJHIHHIIIJJJJJHEHHHFEFFFFFEDEEEEE	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:101	YS:i:-1	YT:Z:CP	NH:i:1
DRR016707.43465443	83	1	19626486	60	100M1S	=	19626483	-104	CATTCTTCTGTAACAATGTTATCAGTAATGCTTTAAACTCCAGCACCTGGTTATGCATTTGAAACCAAGTCTGTTTCTTGTTTTGTATTTTCTCTCTGGAG	DDDDDDCDDDDCCDEEEEDFFFFFDBHHHHHGJJGHDJIJJIGIHJGGJJJJJJJJJJJJJHG?JJJJJJJJJJJIIJJJJIJJJJJIGHGHHFFFFFCCC	AS:i:-1	ZS:i:-12	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:100	YS:i:0	YT:Z:CP	NH:i:1
DRR016707.64407509	147	1	19626484	60	101M	=	19626484	-101	GCCATTCTTCTGTAACAATGTTATCAGTAATGCTTTAAACTCCAGCACCTGGTTATGCATTTGAAACCAAGTCTGTTTCTTGTTTTGTATTTTCTCTCTGG	DDDDDCDDDDDDEEDEEEEFFFFFFHHHHHHJJIJIGHFIHJJJIHFIHHFJJJJJIJJJJJJJIHJJJJJIJJJJJJJJJJJJJJJJHHHHHFFFFFBB@	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:101	YS:i:0	YT:Z:CP	NH:i:1
DRR016707.64407509	99	1	19626484	60	101M	=	19626484	-101	GCCATTCTTCTGTAACAATGTTATCAGTAATGCTTTAAACTCCAGCACCTGGTTATGCATTTGAAACCAAGTCTGTTTCTTGTTTTGTATTTTCTCTCTGG	CCCFFFFFHHHHHJJJJJJJIJJJJJJJJJJJJJJJJJJJJJIJJJJJJIJJDGIJJJJJJJJCGIJJJIIHHJJJHHIJJJHHHHHDFFFFFFEEEEEED	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:101	YS:i:0	YT:Z:CP	NH:i:1</code></pre>
</div>
<p>Field 1: Sequence ID</p>
<p>Field 2: <a href="https://broadinstitute.github.io/picard/explain-flags.html" class="external-link">SAM
Flag</a>. It gives information about the read (is it paired, aligned, is
the mate read properly aligned…)</p>
<p>Field 3: Chromosome where alignment occurs</p>
<p>Field 4: Position of alignment</p>
<p>Field 5: Mapping quality score (read uniquely mapped (60), multiply
mapped (1) or unmapped (0))</p>
<p>Field 6: <a href="https://timd.one/blog/genomics/cigar.php" class="external-link">CIGAR</a>
string. It is a representation of alignment indicating positions of
match/mismatch/deletion/insertion compared to the reference.</p>
<p>Field 7: Reference name of the mate (Set to ‘=’ if the mate’s
reference sequence is the same as this alignment’s)</p>
<p>Field 8: Position of mate</p>
<p>Field 9: Inferred fragment length</p>
<p>Field 10: Read sequence (reverse-complemented if aligned to the
reverse strand)</p>
<p>Field 11: Base quality (Phred score)</p>
<p>Field 12: Optional. See <a href="https://daehwankimlab.github.io/hisat2/manual/" class="external-link">hisat2 manual</a>
for more information.</p>
<p><strong>BAM format</strong></p>
<p>The BAM file is a compressed binary version of SAM file. SAM files
can be easily converted into BAM files using <a href="https://www.htslib.org" class="external-link">samtools</a> software, and it is sometimes
mandatory to sort the SAM/BAM by chromosomal coordinates for downstream
applications.</p>
<pre><code>samtools view -Sb SampleName.sam | samtools sort &gt; SampleName.bam</code></pre>
<p><strong>Alignment visualisation</strong></p>
<p>Alignements described in BAM files can easily be viewed using <a href="https://software.broadinstitute.org/software/igv/" class="external-link">IGV</a>
(Integrative Genomics Viewer), a very useful interactive tool for the
visual exploration of genomic data.</p>
<p>Visualising the alignments is not part of the processing pipeline but
sometimes it can be useful..</p>
<p>IGV requires that BAM files have an associated index file<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>. Samtools
can be used to index the BAM file. The following command would create
the indexed SampleName.bam.bai file.</p>
<pre><code>samtools index SampleName.bam</code></pre>
<p>Bam file can be loaded in IGV<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>. Below is an example of visualisation of a
bam file using IGV, focusing on CDKN1A gene.</p>
<figure><img src="fig/IGV.png" width="100%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Loading a BAM file creates 3 associated tracks:</p>
<ul><li><p>Coverage Track to view depth of coverage</p></li>
<li><p>Splice Junction Track which provides an alternative view of reads
spanning splice junctions</p></li>
<li><p>Alignment Track to view individual aligned reads</p></li>
</ul></section><section><h2 class="section-heading" id="counting">Counting<a class="anchor" aria-label="anchor" href="#counting"></a></h2>
<hr class="half-width"><p>Now comes the quantification step. It can be performed with several
tools such as <a href="https://pypi.org/project/HTSeq/" class="external-link">htseq-count</a>
or <a href="https://subread.sourceforge.net/featureCounts.html" class="external-link">FeatureCounts</a>.
We will describe the latter as it runs much faster. It was developed for
counting reads to genomic features such as genes, exons, promoters and
genomic bins. FeatureCounts takes as input SAM/BAM files and an
annotation file (such as a gtf file) containing chromosomal coordinates
of features.</p>
<p>The Gene Transfer Format (GTF) is a widely used format for storing
gene annotations. Human and mouse GTF files can be downloaded easily
from <a href="https://www.ensembl.org/info/data/ftp/index.html" class="external-link">Ensembl</a> or
from the <a href="https://www.gencodegenes.org/human/" class="external-link">UCSC table
browser</a>. The first few lines of gtf annotation file for human genome
assembly GRCh38 look like this:</p>
<figure><img src="fig/gtf_example.png" width="100%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>FeatureCounts will count the number of uniquely mapped reads assigned
to features (e.g. an exon) or meta-features (e.g. a gene) of the gtf
file. When summarizing reads at gene level, read counts obtained for
exons belonging to the same gene will be added up to yield the read
count for the corresponding gene. Note that an exon-spanning read will
be counted only once for the corresponding gene even if it overlaps with
more than one exon.</p>
<figure><img src="fig/featurecount_metafeature.png" width="80%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>FeatureCounts could be launched with the following command. As usual,
many of other parameters can be fine-tuned (see the <a href="https://manpages.ubuntu.com/manpages/bionic/man1/featureCounts.1.html" class="external-link">manual</a>
for more details).</p>
<pre><code>featureCounts \
-p \                        # paired-end
--countReadPairs \          # fragments counted instead of reads
-a annotations.gtf \        # name of the annotation file
-t exon \                   # feature type in GTF annotation
-g gene_id \                # Meta-features used in GTF annotation
-s 0 \                      # strand-specificity
-o SampleName_counts.tsv \  # Name of the output file
SampleName.bam              # bam file</code></pre>
<p>The featurecounts output file will give us the raw counts of each
gene in that sample. This is how the first lines of featurecounts output
could look like:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           Geneid                   Chr
1 ENSG00000223972     1;1;1;1;1;1;1;1;1
2 ENSG00000227232 1;1;1;1;1;1;1;1;1;1;1
3 ENSG00000278267                     1
4 ENSG00000243485             1;1;1;1;1
5 ENSG00000284332                     1
6 ENSG00000237613             1;1;1;1;1
                                                              Start
1             11869;12010;12179;12613;12613;12975;13221;13221;13453
2 14404;15005;15796;16607;16858;17233;17606;17915;18268;24738;29534
3                                                             17369
4                                     29554;30267;30564;30976;30976
5                                                             30366
6                                     34554;35245;35277;35721;35721
                                                                End
1             12227;12057;12227;12721;12697;13052;13374;14409;13670
2 14501;15038;15947;16765;17055;17368;17742;18061;18366;24891;29570
3                                                             17436
4                                     30039;30667;30667;31109;31097
5                                                             30503
6                                     35174;35481;35481;36073;36081
                 Strand Length sample.bam
1     +;+;+;+;+;+;+;+;+   1735          0
2 -;-;-;-;-;-;-;-;-;-;-   1351         14
3                     -     68          8
4             +;+;+;+;+   1021          0
5                     +    138          0
6             -;-;-;-;-   1219          0</code></pre>
</div>
</section><section><h2 class="section-heading" id="preparing-the-count-matrix">Preparing the count matrix<a class="anchor" aria-label="anchor" href="#preparing-the-count-matrix"></a></h2>
<hr class="half-width"><p>The aim of an RNA-seq experiment is often to compare different types
of cells, or to compare treated cells to control cells, to see if some
genes are differentially expressed.</p>
<p>Once fastq files from each sample have been processed into raw
counts, results can be merged in a single count matrix, were each column
represents a sample, and each line represents a gene. This count matrix
will be the starting point of a differential expression analysis.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/GSE96870_counts_cerebellum.csv"</span>,</span>
<span>                   row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             GSM2545336 GSM2545337 GSM2545338 GSM2545339 GSM2545340 GSM2545341
Xkr4               1891       2410       2159       1980       1977       1945
LOC105243853          0          0          1          4          0          0
LOC105242387        204        121        110        120        172        173
LOC105242467         12          5          5          5          2          6
Rp1                   2          2          0          3          2          1
Sox17               251        239        218        220        261        232
             GSM2545342 GSM2545343 GSM2545344 GSM2545345 GSM2545346 GSM2545347
Xkr4               1757       2235       1779       1528       1644       1585
LOC105243853          1          3          3          0          1          3
LOC105242387        177        130        131        160        180        176
LOC105242467          3          2          2          2          1          2
Rp1                   3          1          1          2          2          2
Sox17               179        296        233        271        205        230
             GSM2545348 GSM2545349 GSM2545350 GSM2545351 GSM2545352 GSM2545353
Xkr4               2275       1881       2584       1837       1890       1910
LOC105243853          1          0          0          1          1          0
LOC105242387        161        154        124        221        272        214
LOC105242467          2          4          7          1          3          1
Rp1                   3          6          5          3          5          1
Sox17               302        286        325        201        267        322
             GSM2545354 GSM2545362 GSM2545363 GSM2545380
Xkr4               1771       2315       1645       1723
LOC105243853          0          1          0          1
LOC105242387        124        189        223        251
LOC105242467          4          2          1          4
Rp1                   3          3          1          0
Sox17               273        197        310        246</code></pre>
</div>
</section><div class="footnotes footnotes-end-of-document">
<hr><ol><li id="fn1"><p>Genome indexing allows the aligner to quickly find
potential alignment sites for query sequences in a genome, which saves
considerable time during alignment. Frequently used genome (human,
mouse, rat…) already have indexed versions that can be downloaded
directly. When working with another reference genome, a specific indexed
genome has to be built.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Indexing a BAM file allows IGV to quickly extract
alignments overlapping particular genomic regions<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>Note that the index file must reside in the same
directory as the bam file that it indexes, and it must have the same
basename<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
</ol></div>



      </div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="08-next-steps.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="10-additionnal-material-LM-FDR.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="08-next-steps.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Next steps</a>
        <a class="chapter-link float-end" href="10-additionnal-material-LM-FDR.html" rel="next">Next: Additional material:... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/09-Additional-material-HTS.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/bioc-rnaseq/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/bioc-rnaseq/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/bioc-rnaseq/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:charlotte.soneson@fmi.ch">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>,
        <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>,
      and <a href="https://github.com/mblue9/varnish/tree/3403ad0dd37f7641c6442330fcc182937e7964c5" class="external-link">varnish (0.2.16)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://UCLouvain-BIOINFO.github.io/bioc-rnaseq/09-Additional-material-HTS.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Additional material: High throughput RNA-sequencing",
  "creativeWorkStatus": "active",
  "url": "https://UCLouvain-BIOINFO.github.io/bioc-rnaseq/09-Additional-material-HTS.html",
  "identifier": "https://UCLouvain-BIOINFO.github.io/bioc-rnaseq/09-Additional-material-HTS.html",
  "dateCreated": "2020-09-15",
  "dateModified": "2025-11-16",
  "datePublished": "2025-12-02"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

