<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-seq analysis with Bioconductor: Additional material: Linear models and FDR</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="manifest" href="../site.webmanifest"><link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link" style="color: #383838">Beta
            <i aria-hidden="true" class="icon" data-feather="alert-circle" style="color: #001483; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
        </abbr>

      </div>
    </div>
    <div class="selector-container ">


      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../093-linear-models-FDR.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-seq analysis with Bioconductor
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-seq analysis with Bioconductor
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html"></a></li>
          </ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq analysis with Bioconductor
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 95%" class="percentage">
    95%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 95%" aria-valuenow="95" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../093-linear-models-FDR.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-intro-to-rnaseq.html">1. Introduction to RNA-seq</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-setup.html">2. RStudio Project and Experimental Data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-import-annotate.html">3. Importing and annotating quantified data into R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-exploratory-qc.html">4. Exploratory analysis and quality control</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-differential-expression.html">5. Differential expression analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-extra-design.html">6. Extra exploration of design matrices</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-gene-set-analysis.html">7. Gene set enrichment analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-next-steps.html">8. Next steps</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-additional.html">9. Additional material</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        10. Additional material: Linear models and FDR
        </span>
      
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush12">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading12">
        <a href="095-pca.html">11. Additional material: PCA</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html"></a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="../instructor/aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/09-additional.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/095-pca.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/09-additional.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Additional material</a>
        <a class="chapter-link float-end" href="../instructor/095-pca.html" rel="next">Next: Additional material:... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Additional material: Linear models and FDR</h1>
        <p> Last updated on 2025-11-10 |

        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/093-linear-models-FDR.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 20 minutes </p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li><p>What are linear models?</p></li>
<li><p>What is a FDR?</p></li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li><p>Explore linear models and learn how to interpret
coefficients</p></li>
<li><p>Address the multiple testing issue</p></li>
</ul></div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="multifactorial-design-and-linear-models">Multifactorial design and linear models<a class="anchor" aria-label="anchor" href="#multifactorial-design-and-linear-models"></a></h1>
<p>The outcome of an experiment, such as a gene expression level in the
case of an RNA-seq, can be seen as the sum of several factors:</p>
<ul><li><p>the treatments effects (these are the effects you actually want
to study)</p></li>
<li><p>biological effects (representing natural biological differences
between samples that you might also have to consider)</p></li>
<li><p>technical effects (that come from the experimental process itself
and that could potentially influence the measurements)</p></li>
<li><p>and lastly, outcome will also reflect an error, which represent
the random variability (or experiment noise) that cannot be explain be
any of the known factors.</p></li>
</ul><figure><img src="../fig/factors-affecting-measurements.png" alt="A classification of many different factors affecting measurements obtained from an experiment into treatment, biological, technical and error effects" class="figure mx-auto d-block"></figure><p>(figure from <a href="https://www.amazon.com/Experimental-Design-Laboratory-Biologists-Reproducibility/dp/1107424887" class="external-link">Lazic
(2017)</a>).</p>
<div class="section level2">
<h2 id="single-factor-design">Single-factor design<a class="anchor" aria-label="anchor" href="#single-factor-design"></a></h2>
<p>Let’s start by a simple design, and let’s say we have the gene
expression values of 3 control samples and 3 samples treated with a
drug, assuming that there is no other biological effect or batch effect
to consider.</p>
<p>We can use the following linear model to modelise the expression
values:</p>
<p><span class="math display">\[y = \beta_0 + \beta_1 . x\]</span></p>
<p>In this equation,</p>
<ul><li><p><span class="math inline">\(y\)</span> would be the measured
expression level of a gene</p></li>
<li><p>the design factor <span class="math inline">\(x\)</span> is a
binary indicator variable representing the two groups of our experiment.
It will take the value of 0 if we are considering the control samples,
and a value of 1 when considering treated samples.</p></li>
</ul><p><strong>When considering a control sample</strong>, the <span class="math inline">\(x\)</span> factor is set to 0.</p>
<p>The <span class="math inline">\(\beta_1 . x\)</span> term is hence
equal to 0, so the equation can be written:</p>
<p><span class="math inline">\(y = \beta_0\)</span></p>
<p>This means that the <span class="math inline">\(\beta_0\)</span> term
represents the expression level of the gene in the control samples, this
coefficient is usually called the <strong>intercept</strong></p>
<p><strong>When considering a treated sample</strong>, the <span class="math inline">\(x\)</span> factor is set to 1.</p>
<p>The equation becomes:</p>
<p><span class="math inline">\(y = \beta_0 + \beta_1\)</span></p>
<p>Assuming our expression values are in a log scale then</p>
<p><span class="math inline">\(\beta_1 = y - \beta_0 =
log(expression_{Treated\_cells}) -
log(expression_{Control\_cells})\)</span></p>
<p><span class="math inline">\(\beta_1 =
log(\frac{expression_{Treated\_cells}}{expression_{Control\_cells}})\)</span></p>
<p>The <span class="math inline">\(\beta_1\)</span> term represents
hence the <strong>logFoldchange</strong> between the treated and the
control samples (the treatment effect).</p>
<p><strong>The goal is now to estimate the coefficients and test their
significance</strong></p>
<p>The <span class="math inline">\(\beta_0\)</span> and <span class="math inline">\(\beta_1\)</span> coefficients can be estimated
from the data using the expression values in the control and in the
treated samples.</p>
<p>A statistical test can then be applied, to test if the <span class="math inline">\(\beta_1\)</span> coefficient (the logFC due to the
treatment effect) is significantly different from 0.</p>
<p>The null hypothesis is that there is no difference across the sample
groups, which is the same as saying that the <span class="math inline">\(\beta_1\)</span> = 0.</p>
<ul><li>
<span class="math inline">\(H_0\)</span>: <span class="math inline">\(\beta_1 = 0\)</span> (the logFC = 0)</li>
<li>
<span class="math inline">\(H_1\)</span>: <span class="math inline">\(\beta_1 \neq 0\)</span> (the logFC )</li>
</ul><div class="section level3">
<h3 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a></h3>
<p>We are going to use as a toy example a subsetted version of the rna
dataset. Let’s focus only on the first gene, the gene <em>Asl</em>, and
let’s subset the table to keep only female mice at time 0 and 8.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/rnaseq.csv"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Rows: 32428 Columns: 19
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (14): gene, sample, organism, sex, infection, strain, tissue, product, e...
dbl  (5): expression, age, time, mouse, ENTREZID

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna1</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">gene</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Asl"</span><span class="op">)</span>, <span class="va">time</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">8</span><span class="op">)</span>, <span class="va">sex</span> <span class="op">==</span> <span class="st">"Female"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>expression_log <span class="op">=</span> <span class="fu">log</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">expression_log</span>, <span class="va">time</span>, <span class="va">infection</span>, <span class="va">sex</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">gene</span>, <span class="va">infection</span><span class="op">)</span> </span></code></pre>
</div>
<p>Let’s also convert the <code>infection</code> column into a factor,
in order to set the <code>NonInfected</code> condition as the reference
level.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna1</span><span class="op">$</span><span class="va">infection</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">rna1</span><span class="op">$</span><span class="va">infection</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"NonInfected"</span>, <span class="st">"InfluenzaA"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Assuming the expression values follow a normal distribution (even
though, as we’ll see later, this isn’t actually true!) and let’s use the
<code>lm()</code> function to fit a linear model to compare infected
mice to non-infected mice.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod1</span> <span class="op">&lt;-</span> <span class="fu">lm</span><span class="op">(</span><span class="va">expression_log</span> <span class="op">~</span> <span class="va">infection</span>, data <span class="op">=</span> <span class="va">rna1</span><span class="op">)</span></span>
<span><span class="va">mod1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Call:
lm(formula = expression_log ~ infection, data = rna1)

Coefficients:
        (Intercept)  infectionInfluenzaA
             6.0941               0.8358  </code></pre>
</div>
<p>The lm() function has estimated the coefficients of the linear
model:</p>
<ul><li><p>The estimate for the <span class="math inline">\(\beta_0\)</span>
coefficient (the <code>Intercept</code>) is 6.0941</p></li>
<li><p>The estimate for the <span class="math inline">\(\beta_1\)</span>
coefficient (annotated as the <code>infectionInfluenzaA</code>
coefficient) is 0.8358</p></li>
</ul><p>We can visualise this linear model with:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna1</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">infection</span>, y <span class="op">=</span> <span class="va">expression_log</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"NonInfected"</span>, y <span class="op">=</span> <span class="fl">6.094074</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"blue"</span>, size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"InfluenzaA"</span>, y <span class="op">=</span> <span class="fl">6.094074</span> <span class="op">+</span> <span class="fl">0.8357904</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"red"</span>, size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/093-linear-models-FDR-rendered-unnamed-chunk-5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>And indeed, we see that</p>
<p>– the intercept (the <span class="math inline">\(\beta_0\)</span>
coefficient) corresponds to the mean expression level of the gene in the
<code>Noninfected</code> group corresponds (the blue dot)</p>
<p>– the sum of the <span class="math inline">\(\beta_0\)</span> and the
<span class="math inline">\(\beta_1\)</span> coefficients (the intercept
+ the logFC) corresponds to the mean expression level of the gene in the
<code>InfluenzaA</code> group (the red dot)</p>
<p>We can also extract the p-value associated with the
<code>infectionInfluenzaA</code> coefficient:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">mod1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Call:
lm(formula = expression_log ~ infection, data = rna1)

Residuals:
     Min       1Q   Median       3Q      Max
-0.20520 -0.12727  0.01064  0.13955  0.19934

Coefficients:
                    Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)          6.09407    0.08991  67.782 6.94e-10 ***
infectionInfluenzaA  0.83579    0.12715   6.573 0.000594 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.1798 on 6 degrees of freedom
Multiple R-squared:  0.8781,	Adjusted R-squared:  0.8578
F-statistic: 43.21 on 1 and 6 DF,  p-value: 0.0005944</code></pre>
</div>
<p>We can see that the coefficient <code>infectionInfluenzaA</code> is
highly significant.</p>
</div>
</div>
<div class="section level2">
<h2 id="two-factor-design">Two-factor design<a class="anchor" aria-label="anchor" href="#two-factor-design"></a></h2>
<p>Let’s consider a slightly more complex design. We now have the gene
expression values of 6 control and 6 treated samples, but the experiment
was conducted in 2 different cell types (cells A and cells B). We hence
have 3 controls and 3 treated samples for each cell type.</p>
<p>To take into account the treatment effect while controlling the cell
origin, we can use the following linear model:</p>
<p><span class="math display">\[y = \beta_0 + \beta_1 . x_1 + \beta_2 .
x_2\]</span></p>
<p>We could rewrite this equation as:</p>
<p><span class="math display">\[y = \beta_0 + Treatment. x_1 + Cell .
x_2\]</span></p>
<p>In this equation,</p>
<ul><li><p>the <span class="math inline">\(\beta_0\)</span> term represents
the <code>intercept</code>, i.e, the <strong>expression level of the
gene in the reference sample</strong>, for instance the untreated cell
type A.</p></li>
<li><p>the design factor <span class="math inline">\(x_1\)</span> is a
binary indicator variable representing the two treatment conditions of
our experiment. It will take the value of 0 if we are considering the
untreated samples, and a value of 1 when considering treated
samples.</p></li>
<li><p>The <code>Treatment</code> term represents the
<strong>logFoldchange due to treatment</strong>.</p></li>
<li><p>the design factor <span class="math inline">\(x_2\)</span> is
another binary indicator variable representing the two cell types of our
experiment. It will take the value of 0 if we are considering the cell
type A, and a value of 1 when considering cell type B.</p></li>
<li><p>The <code>Cell</code> term represents the <strong>logFoldchange
due to celltype</strong>.</p></li>
</ul><p>The following figure (generated with the
<code>ExploreModelMatrix</code> package) illustrates the value of the
linear predictor of a linear model for each combination of input
variables.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="../fig/093-linear-models-FDR-rendered-unnamed-chunk-7-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>From this figure, we can see how the genes log expression values are
modelised in the different samples:</p>
<ul><li><p>In <code>untreated cells A</code>:          <span class="math inline">\(y\)</span> = <span class="math inline">\(\beta_0\)</span></p></li>
<li><p>In <code>treated cells A</code>:              <span class="math inline">\(y\)</span> = <span class="math inline">\(\beta_0\)</span> + <em>Treatment</em></p></li>
<li><p>In <code>untreated cells B</code>:          <span class="math inline">\(y\)</span> = <span class="math inline">\(\beta_0\)</span> + <em>Cell</em></p></li>
<li><p>In <code>treated cells B</code>:              <span class="math inline">\(y\)</span> = <span class="math inline">\(\beta_0\)</span> + <em>Treatment</em> +
<em>Cell</em></p></li>
</ul><div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge1" class="callout-inner">
<h3 class="callout-title">Challenge1:</h3>
<div class="callout-content">
<p>Subset the rna dataset as below to keep only on the gene
<em>Ddx3x</em>, and mice at time 0 and 8.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna2</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">gene</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Ddx3x"</span><span class="op">)</span>, <span class="va">time</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">8</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>expression_log <span class="op">=</span> <span class="fu">log</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">expression_log</span>, <span class="va">time</span>, <span class="va">infection</span>, <span class="va">sex</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">gene</span>, <span class="va">infection</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">#set the `NonInfected` condition as the reference level. </span></span>
<span><span class="va">rna2</span><span class="op">$</span><span class="va">infection</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">rna2</span><span class="op">$</span><span class="va">infection</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"NonInfected"</span>, <span class="st">"InfluenzaA"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Use the <code>lm()</code> function to fit a linear model that would
account for infection and sex.</p>
<p>Interprete the results.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod2</span> <span class="op">&lt;-</span> <span class="fu">lm</span><span class="op">(</span><span class="va">expression_log</span> <span class="op">~</span> <span class="va">infection</span> <span class="op">+</span> <span class="va">sex</span>, data <span class="op">=</span> <span class="va">rna2</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">mod2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Call:
lm(formula = expression_log ~ infection + sex, data = rna2)

Residuals:
      Min        1Q    Median        3Q       Max
-0.163831 -0.019300  0.005398  0.030459  0.076010

Coefficients:
                    Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)         10.06641    0.02789 360.888  &lt; 2e-16 ***
infectionInfluenzaA  0.15182    0.03364   4.513 0.000882 ***
sexMale             -0.28277    0.03399  -8.319 4.49e-06 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.06294 on 11 degrees of freedom
Multiple R-squared:  0.8906,	Adjusted R-squared:  0.8707
F-statistic: 44.79 on 2 and 11 DF,  p-value: 5.175e-06</code></pre>
</div>
<p>The coefficients <code>infectionInfluenzaA</code> (representing the
<em>Infection Effect</em>) and the coefficient <code>sexMale</code>,
representing the <em>Sex Effect</em>, are both significant.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna2</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">infection</span>, y <span class="op">=</span> <span class="va">expression_log</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span>, <span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">sex</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/093-linear-models-FDR-rendered-unnamed-chunk-10-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
</div>
</div>
</div>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge2" class="callout-inner">
<h3 class="callout-title">Challenge2:</h3>
<div class="callout-content">
<p>Re-use the <code>lm()</code> function to fit a linear model that
would only account for infection.</p>
<p>Compare with the previous model and try to interprete the
difference.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod3</span> <span class="op">&lt;-</span> <span class="fu">lm</span><span class="op">(</span><span class="va">expression_log</span> <span class="op">~</span> <span class="va">infection</span>, data <span class="op">=</span> <span class="va">rna2</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">mod3</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Call:
lm(formula = expression_log ~ infection, data = rna2)

Residuals:
     Min       1Q   Median       3Q      Max
-0.20156 -0.15940  0.02542  0.14115  0.19720

Coefficients:
                    Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)          9.94523    0.06150 161.710   &lt;2e-16 ***
infectionInfluenzaA  0.15182    0.08697   1.746    0.106
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.1627 on 12 degrees of freedom
Multiple R-squared:  0.2025,	Adjusted R-squared:  0.136
F-statistic: 3.047 on 1 and 12 DF,  p-value: 0.1064</code></pre>
</div>
<p>The <em>Infection Effect</em> is not significant anymore.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna2</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">infection</span>, y <span class="op">=</span> <span class="va">expression_log</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span>, <span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/093-linear-models-FDR-rendered-unnamed-chunk-12-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="design-with-interaction">Design with interaction<a class="anchor" aria-label="anchor" href="#design-with-interaction"></a></h2>
<p>Coming back to the previous experimental design were we considered 3
controls and 3 treated samples for each cell type (cells A and cells B).
An interaction term could be added to the linear model, to modelise a
treatment effect that would differ across the cell types.</p>
<p>The linear model could be written</p>
<p><span class="math display">\[y = \beta_0 + \beta_1 . x_1 + \beta_2 .
x_2 + \beta_{12}.x_1.x_2.\]</span></p>
<p>We could rewrite this equation as:</p>
<p><span class="math display">\[y = \beta_0 + Treatment. x_1 + Cell .
x_2 + Treatment.Cell.x_1.x_2\]</span></p>
<p>In this equation,</p>
<ul><li><p>the <span class="math inline">\(\beta_0\)</span> term represents
the <code>intercept</code>, i.e, the <strong>expression level of the
gene in the reference sample</strong>, for instance the untreated cell
type A.</p></li>
<li><p>the design factor <span class="math inline">\(x_1\)</span> will
take the value of 0 if we are considering the untreated samples, and a
value of 1 when considering treated samples.</p></li>
<li><p>The <code>Treatment</code> term represents the <strong>treatment
effect in the reference cell type</strong>. In this case it would
represent the logFoldchange between treated and untreated celltype
A.</p></li>
<li><p>the design factor <span class="math inline">\(x_2\)</span> will
take the value of 0 if we are considering the cell type A, and a value
of 1 when considering cell type B.</p></li>
<li><p>The <code>Cell</code> term represents the <strong>Celltype effect
in the untreated cells</strong>. In this case it would represent the
logFoldchange between untreated celltype A and untreated celltype
A.</p></li>
<li><p>The <code>Treatment.Cell</code> term represents the
<strong>treatment effect that would be different in cell type B compared
to cell type A</strong>. This term will only be considered in the
equation when we are referring to treated celltype B (i.e. when <span class="math inline">\(x_1\)</span> = 1 and <span class="math inline">\(x_2\)</span> = 1)</p></li>
</ul><p>The following figure (generated with the
<code>ExploreModelMatrix</code> package) illustrates the value of the
linear predictor of a linear model for each combination of input
variables.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="../fig/093-linear-models-FDR-rendered-unnamed-chunk-13-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>From this figure, we can see how the genes log expression values are
modelised in the different samples:</p>
<ul><li><p>In <code>untreated cells A</code>:          <span class="math inline">\(y\)</span> = <span class="math inline">\(\beta_0\)</span></p></li>
<li><p>In <code>treated cells A</code>:              <span class="math inline">\(y\)</span> = <span class="math inline">\(\beta_0\)</span> + <em>Treatment</em></p></li>
<li><p>In <code>untreated cells B</code>:          <span class="math inline">\(y\)</span> = <span class="math inline">\(\beta_0\)</span> + <em>Cell</em></p></li>
<li><p>In <code>treated cells B</code>:              <span class="math inline">\(y\)</span> = <span class="math inline">\(\beta_0\)</span> + <em>Treatment</em> +
<em>Cell</em> + <em>Treatment.Cell</em></p></li>
</ul><div id="challenge1-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge1-1" class="callout-inner">
<h3 class="callout-title">Challenge1:</h3>
<div class="callout-content">
<p>Subset the rna dataset as below to keep only on the gene
<em>Ddx3x</em>, and mice at time 0 and 8.</p>
<p>Re-use the previous <code>rna2</code> dataset and fit a linear model
with an interaction between infection and sex.</p>
<p>Interprete the results.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod3</span> <span class="op">&lt;-</span> <span class="fu">lm</span><span class="op">(</span><span class="va">expression_log</span> <span class="op">~</span> <span class="va">infection</span> <span class="op">*</span> <span class="va">sex</span>, data <span class="op">=</span> <span class="va">rna2</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">mod3</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Call:
lm(formula = expression_log ~ infection * sex, data = rna2)

Residuals:
      Min        1Q    Median        3Q       Max
-0.155912 -0.025986  0.005398  0.032198  0.083929

Coefficients:
                            Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)                 10.07433    0.03256 309.431  &lt; 2e-16 ***
infectionInfluenzaA          0.13598    0.04604   2.953 0.014451 *
sexMale                     -0.30125    0.04973  -6.057 0.000122 ***
infectionInfluenzaA:sexMale  0.03696    0.07033   0.525 0.610714
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.06512 on 10 degrees of freedom
Multiple R-squared:  0.8936,	Adjusted R-squared:  0.8616
F-statistic: 27.99 on 3 and 10 DF,  p-value: 3.529e-05</code></pre>
</div>
<p>The coefficients <code>infectionInfluenzaA</code> (representing the
<em>Infection Effect</em>) and the coefficient <code>sexMale</code>,
representing the <em>Sex Effect</em>, are both significant.</p>
<p>But the interaction term is not. This indicates that the
<em>Infection Effect</em> is not significantly different between males
and females</p>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="multiple-testing-issue">multiple testing issue<a class="anchor" aria-label="anchor" href="#multiple-testing-issue"></a></h1>
<p>In the few examples of linear model that we have seen in the previous
section, the <code>lm()</code> function was always applied to a single
gene.</p>
<p>In a real RNA-seq experiment, we will have a few thousand tests to
perform, as each gene is going to be tested</p>
<p>To introduce the multiple testing issue, let’s refer to <a href="https://xkcd.com/882/" class="external-link">this xkcd cartoon</a> that depicts
scientists testing whether eating jelly beans causes acne.</p>
<figure style="text-align: center"><img src=".././fig/jellybeans.png" alt="Do jelly beans cause acne? Scientists investigate. From [xkcd](https://xkcd.com/882/)." width="635" class="figure mx-auto d-block"><figcaption>
Do jelly beans cause acne? Scientists investigate. From <a href="https://xkcd.com/882/" class="external-link">xkcd</a>.
</figcaption></figure><p>Let’s illustrate the multiple testing problem using simulated
expression data. We are going to create a simulated expression matrix
containing expression values for 20,000 genes, all generated from the
same distribution <span class="math inline">\(N(100, 20)\)</span>.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">120000</span>, mean <span class="op">=</span> <span class="fl">100</span>, sd <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">exp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20000</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">exp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"A"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span>, <span class="fu">paste0</span><span class="op">(</span><span class="st">"B"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">exp</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             A1        A2        A3       B1        B2        B3
Gene1  76.40680  80.14777  80.89500 83.08430  61.52934 132.58000
Gene2  95.05106 103.21809 108.28011 98.85551 118.65169  54.78860
Gene3  95.82976 109.70075  99.40699 89.99937 124.73202 114.73639
Gene4  76.17949 106.12586 121.83424 79.34932  91.80822 113.07296
Gene5 118.46590 102.40266 101.35891 96.24010  88.27471 102.95286
Gene6  71.92304  91.23588 104.79177 98.89892 103.32194  85.51648</code></pre>
</div>
<p>The data is normally distributed so let’s apply a t.test on one of
these genes</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="fu">t.test</span><span class="op">(</span><span class="va">exp</span><span class="op">[</span><span class="va">n</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>, <span class="va">exp</span><span class="op">[</span><span class="va">n</span>, <span class="fl">4</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
	Welch Two Sample t-test

data:  exp[n, 1:3] and exp[n, 4:6]
t = -0.62851, df = 2.0174, p-value = 0.5934
alternative hypothesis: true difference in means is not equal to 0
95 percent confidence interval:
 -103.19405   76.69801
sample estimates:
mean of x mean of y
 79.14986  92.39788 </code></pre>
</div>
<p>Now, let’s apply the t.test on all genes</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp_tbl</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">exp</span>, rownames <span class="op">=</span> <span class="st">"Gene"</span><span class="op">)</span> </span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="va">exp_tbl</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">rowwise</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>pval <span class="op">=</span> <span class="fu">t.test</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">A1</span>, <span class="va">A2</span>, <span class="va">A3</span><span class="op">)</span>, <span class="fu">c</span><span class="op">(</span><span class="va">B1</span>,<span class="va">B2</span>,<span class="va">B3</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span> <span class="co"># to revert the rowwise()</span></span>
<span><span class="va">res</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 20,000 × 8
   Gene      A1    A2    A3    B1    B2    B3  pval
   &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 Gene1   76.4  80.1  80.9  83.1  61.5 133.  0.593
 2 Gene2   95.1 103.  108.   98.9 119.   54.8 0.609
 3 Gene3   95.8 110.   99.4  90.0 125.  115.  0.522
 4 Gene4   76.2 106.  122.   79.3  91.8 113.  0.712
 5 Gene5  118.  102.  101.   96.2  88.3 103.  0.177
 6 Gene6   71.9  91.2 105.   98.9 103.   85.5 0.587
 7 Gene7  100.   68.6  86.9  81.3 148.   88.7 0.438
 8 Gene8   86.6 106.   95.8 111.   89.5 108.  0.486
 9 Gene9   77.3  90.3  84.0  69.6 107.   97.1 0.589
10 Gene10 142.   96.6 168.  146.  126.   94.4 0.632
# ℹ 19,990 more rows</code></pre>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<ul><li><p>How many significant genes did we obtained ? Was it
expected?</p></li>
<li><p>Why do we obtain genes with a p-value &lt; 0.05?</p></li>
</ul></div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
FALSE  TRUE
19271   729 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sign_genes</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> </span>
<span><span class="va">unsign_genes</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">pval</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">res_long</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"sample"</span>, values_to <span class="op">=</span> <span class="st">"expression"</span>, <span class="op">-</span><span class="fu">c</span><span class="op">(</span><span class="va">Gene</span>, <span class="va">pval</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>group <span class="op">=</span> <span class="fu">substr</span><span class="op">(</span><span class="va">sample</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">fig1</span> <span class="op">&lt;-</span> <span class="va">res_long</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">expression</span>, fill <span class="op">=</span> <span class="va">group</span>, color <span class="op">=</span> <span class="va">group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_density</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>data <span class="op">=</span> <span class="va">res_long</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">unsign_genes</span><span class="op">$</span><span class="va">Gene</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,</span>
<span>             <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">expression</span>, y <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="va">group</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Gene with a pval &gt; 0.05"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">fig2</span> <span class="op">&lt;-</span> <span class="va">res_long</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">expression</span>, fill <span class="op">=</span> <span class="va">group</span>, color <span class="op">=</span> <span class="va">group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_density</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>data <span class="op">=</span> <span class="va">res_long</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">Gene</span> <span class="op">==</span> <span class="va">sign_genes</span><span class="op">$</span><span class="va">Gene</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,</span>
<span>             <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">expression</span>, y <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="va">group</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Gene with a pval &lt; 0.05"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fig1</span> <span class="op">/</span> <span class="va">fig2</span></span></code></pre>
</div>
<figure><img src="../fig/093-linear-models-FDR-rendered-unnamed-chunk-19-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
</div>
</div>
</div>
<div id="challenge2-discuss" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge2-discuss" class="callout-inner">
<h3 class="callout-title">Challenge2: Discuss</h3>
<div class="callout-content">
<ul><li>If you were to visualize the p-value distribution using a histogram,
how would you expect it to look like?</li>
</ul></div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">pval</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">0.05</span>, boundary <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"gray"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/093-linear-models-FDR-rendered-unnamed-chunk-20-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
</div>
</div>
</div>
<p>Now, let’s slightly modify our simulated expression data.</p>
<p>In a typical RNA-seq experiment, we may test around 20,000 genes, but
only a fraction of them are usually truly differentially expressed. We
could imagine a drug treatment that would modify the expression of about
1000 genes, but that would have no impact on the other ones.</p>
<p>To mimic this, we are going to simulate an up-regulation for the
first 1,000 genes by increasing their expression values in group B.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp_modified</span> <span class="op">&lt;-</span> <span class="va">exp</span></span>
<span></span>
<span><span class="va">exp_modified</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span>, <span class="fl">4</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="fl">1000</span><span class="op">*</span><span class="fl">3</span>, mean <span class="op">=</span> <span class="fl">300</span>, sd <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">exp_modified_tbl</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">exp_modified</span>, rownames <span class="op">=</span> <span class="st">"Gene"</span><span class="op">)</span> </span>
<span><span class="va">exp_modified_tbl</span><span class="op">$</span><span class="va">DE</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="cn">TRUE</span>, <span class="fl">1000</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="cn">FALSE</span>, <span class="fl">19000</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">exp_modified_tbl</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"sample"</span>, values_to <span class="op">=</span> <span class="st">"expression"</span>, <span class="op">-</span><span class="fu">c</span><span class="op">(</span><span class="va">Gene</span>, <span class="va">DE</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>group <span class="op">=</span> <span class="fu">substr</span><span class="op">(</span><span class="va">sample</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">expression</span>, fill <span class="op">=</span> <span class="va">group</span>, color <span class="op">=</span> <span class="va">group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_density</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">.3</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/093-linear-models-FDR-rendered-unnamed-chunk-21-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Let’s run a t-test on this new dataset:</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_modified</span> <span class="op">&lt;-</span> <span class="va">exp_modified_tbl</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">rowwise</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>pval <span class="op">=</span> <span class="fu">t.test</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">A1</span>, <span class="va">A2</span>, <span class="va">A3</span><span class="op">)</span>, <span class="fu">c</span><span class="op">(</span><span class="va">B1</span>, <span class="va">B2</span>, <span class="va">B3</span><span class="op">)</span>, var <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">table</span><span class="op">(</span><span class="va">res_modified</span><span class="op">$</span><span class="va">pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
FALSE  TRUE
18002  1998 </code></pre>
</div>
<div id="challenge2-discuss-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge2-discuss-1" class="callout-inner">
<h3 class="callout-title">Challenge2: Discuss</h3>
<div class="callout-content">
<ul><li><p>How do think the histogram of the p-values of the first 1000
would look like?</p></li>
<li><p>What about the histogram of all p-values?</p></li>
</ul></div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h1</span> <span class="op">&lt;-</span> <span class="va">res_modified</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">DE</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">pval</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">0.05</span>, boundary <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">xlim</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ylim</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1500</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Ho is FALSE"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">h2</span> <span class="op">&lt;-</span> <span class="va">res_modified</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="op">!</span><span class="va">DE</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">pval</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">0.05</span>, boundary <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">xlim</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ylim</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1500</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Ho is TRUE"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">h3</span> <span class="op">&lt;-</span> <span class="va">res_modified</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">pval</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_histogram</span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">0.05</span>, boundary <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">950</span>, color <span class="op">=</span> <span class="st">"red"</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"All genes"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">h1</span> <span class="op">+</span> <span class="va">h2</span> <span class="op">+</span> <span class="va">h3</span></span></code></pre>
</div>
<figure><img src="../fig/093-linear-models-FDR-rendered-unnamed-chunk-23-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The figure on the right illustrates the principle behind the FDR
adjustment. The False discovery rate (FDR) is the expected proportion of
false positives among all the results considered as “significant”
(i.e. pval &lt; 0.05). The adjustment procedure estimate the proportion
of hypothesis that are null and then adjust the p-values
accordingly.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="adjustment-for-multiple-testing">Adjustment for multiple testing<a class="anchor" aria-label="anchor" href="#adjustment-for-multiple-testing"></a></h3>
<p>The False discovery rate (FDR) computes the expected fraction of
false positives among all discoveries. It allows us to choose n results
with a given FDR. Widely used examples are Benjamini-Hochberg or
q-values.</p>
<p>let’s use the <code>p.adjust()</code> function to adjust the
pvalues.</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_modified</span> <span class="op">&lt;-</span> <span class="va">res_modified</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>padj <span class="op">=</span> <span class="fu">p.adjust</span><span class="op">(</span><span class="va">pval</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu">table</span><span class="op">(</span>sign_padj <span class="op">=</span> <span class="va">res_modified</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, sign_pval <span class="op">=</span> <span class="va">res_modified</span><span class="op">$</span><span class="va">pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>         sign_pval
sign_padj FALSE  TRUE
    FALSE 18002   970
    TRUE      0  1028</code></pre>
</div>
<p>Only 1028 were found to be significant after FDR correction.</p>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<ul><li>How many genes would you expect to remain significant after FDR
correction in the first dataset?</li>
</ul></div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7"> Show me the solution </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" data-bs-parent="#accordionSolution7" aria-labelledby="headingSolution7">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">mutate</span><span class="op">(</span>padj <span class="op">=</span> <span class="fu">p.adjust</span><span class="op">(</span><span class="va">pval</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">table</span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pval</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
FALSE  TRUE
19271   729 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span></span>
<span><span class="cn">FALSE</span></span>
<span><span class="fl">20000</span> </span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>



      </div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/09-additional.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/095-pca.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/09-additional.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Additional material</a>
        <a class="chapter-link float-end" href="../instructor/095-pca.html" rel="next">Next: Additional material:... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/093-linear-models-FDR.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/bioc-rnaseq/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/bioc-rnaseq/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/bioc-rnaseq/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:charlotte.soneson@fmi.ch">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="../LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>,
        <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>,
      and <a href="https://github.com/mblue9/varnish/tree/3403ad0dd37f7641c6442330fcc182937e7964c5" class="external-link">varnish (0.2.16)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://UCLouvain-BIOINFO.github.io/bioc-rnaseq/instructor/093-linear-models-FDR.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Additional material: Linear models and FDR",
  "creativeWorkStatus": "active",
  "url": "https://UCLouvain-BIOINFO.github.io/bioc-rnaseq/instructor/093-linear-models-FDR.html",
  "identifier": "https://UCLouvain-BIOINFO.github.io/bioc-rnaseq/instructor/093-linear-models-FDR.html",
  "dateCreated": "2020-09-15",
  "dateModified": "2025-11-10",
  "datePublished": "2025-11-11"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

