<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-seq analysis with Bioconductor: Additional material: GSEA</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="manifest" href="../site.webmanifest"><link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link" style="color: #383838">Beta
            <i aria-hidden="true" class="icon" data-feather="alert-circle" style="color: #001483; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
        </abbr>

      </div>
    </div>
    <div class="selector-container ">


      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../12-GSEA.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-seq analysis with Bioconductor
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-seq analysis with Bioconductor
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html"></a></li>
          </ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq analysis with Bioconductor
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 95%" class="percentage">
    95%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 95%" aria-valuenow="95" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../12-GSEA.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-intro-to-rnaseq.html">1. Introduction to RNA-seq</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-setup.html">2. RStudio Project and Experimental Data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-import-annotate.html">3. Importing and annotating quantified data into R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-exploratory-qc.html">4. Exploratory analysis and quality control</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-differential-expression.html">5. Differential expression analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-extra-design.html">6. Extra exploration of design matrices</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-gene-set-analysis.html">7. Gene set enrichment analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-next-steps.html">8. Next steps</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-Additional-material-HTS.html">9. Additional material: High throughput RNA-sequencing</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-additionnal-material-LM-FDR.html">10. Additional material: Linear models and FDR</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush12">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading12">
        <a href="11-Additional-material-PCA.html">11. Additional material: PCA</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        12. Additional material: GSEA
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#objectives-of-an-enrichment-analysis">Objectives of an enrichment analysis</a></li>
<li><a href="#genesets">GeneSets</a></li>
<li><a href="#input-data">Input data</a></li>
<li><a href="#ora">ORA</a></li>
<li><a href="#gsea">GSEA</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html"></a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="../instructor/aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/11-Additional-material-PCA.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/11-Additional-material-PCA.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Additional material:</a>
        <a class="chapter-link float-end" href="../instructor/index.html" rel="next">Next... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Additional material: GSEA</h1>
        <p> Last updated on 2025-11-16 |

        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/12-GSEA.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 40 minutes </p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What is the aim of performing gene set enrichment analysis?</li>
<li>What are the commonly-used gene set databases?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Learn how to obtain gene sets from various resources in R.</li>
<li>Learn how to perform ORA and GSEA analyses</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="objectives-of-an-enrichment-analysis">Objectives of an enrichment analysis<a class="anchor" aria-label="anchor" href="#objectives-of-an-enrichment-analysis"></a></h2>
<hr class="half-width"><p>Once we have obtained a list of differentially expressed (DE) genes,
the next question naturally to ask is what biological functions these DE
genes may affect.</p>
<p>Gene set enrichment analyses (GSEA) evaluate the associations of a
list of DE genes to a collection of pre-defined gene sets, where each
gene set has a specific biological meaning. A geneset significantly
enrichiched among DE genes could suggest that the corresponding
biological process or pathway is significantly affected.</p>
<p>There are a huge amount of methods available for GSEA analysis. In
this episode, we will focus on two widely used methods: the
over-representation analysis (ORA) and the Gene set enrichment analysis
(GSEA).</p>
</section><section><h2 class="section-heading" id="genesets">GeneSets<a class="anchor" aria-label="anchor" href="#genesets"></a></h2>
<hr class="half-width"><p>The definition of a gene set is very flexible and the construction of
gene sets is straightforward. In most cases, gene sets are from public
databases where huge efforts from scientific curators have already been
made to carefully categorize genes into gene sets with clear biological
meanings. For example, in a “cell cycle” gene set, all the genes are
involved in the cell cycle process. Thus, if DE genes are significantly
enriched in the “cell cycle” gene set, which means there are
significantly more cell cycle genes differentially expressed than
expected, we can conclude that the normal function of cell cycle process
may be affected.</p>
<p>Nevertheless, gene sets can also be self-defined from individual
studies, such as a set of genes in a network module from a co-expression
network analysis, or a set of genes that are up-regulated in a certain
disease.</p>
<p>The <a href="https://www.gsea-msigdb.org/gsea/msigdb" class="external-link">MSigDB
database</a> contains a collection of annotated gene sets such as</p>
<ul><li><code>Gene Ontology (GO)</code></li>
</ul><p>Gene Ontology defines concepts/classes used to describe gene
function, and relationships between these concepts. GO terms are
organized in a directed acyclic graph, where edges between terms
represent parent-child relationship. It classifies functions along three
aspects:</p>
<p>– <code>MF: Molecular Function</code>: molecular activities of gene
products</p>
<p>– <code>CC: Cellular Component</code>: where gene products are
active</p>
<p>– <code>BP: Biological Process</code> pathways and larger processes
made up of the activities of multiple gene products.</p>
<ul><li><p><code>Hallmark gene sets</code> are curated gene sets that
represent well-defined biological states or processes (e.g. “Apoptosis”,
“KRAS signaling”, “DNA repair”).</p></li>
<li><p><code>Kyoto Encyclopedia of Genes and Genomes (KEGG)</code></p></li>
</ul><p>KEGG is a collection of manually drawn pathway maps representing
molecular interaction and reaction networks. These pathways cover a wide
range of biochemical processes.</p>
<ul><li><code>Other gene sets</code></li>
</ul><p>Other gene sets include but are not limited to Positional gene sets,
wikiPathways, Oncogenic signatures, Immunologic signatures…</p>
<div class="section level3">
<h3 id="access-the-gene-sets-from-msigdb">Access the gene sets from MSigDB<a class="anchor" aria-label="anchor" href="#access-the-gene-sets-from-msigdb"></a></h3>
<p>The <a href="https://cran.r-project.org/web/packages/msigdbr/index.html" class="external-link">msigdbr</a>
CRAN package provides the MSigDB gene sets in a standard R data frame
with key-value pairs.</p>
<p>One can use msigdbr_collections() to see the available
collections</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">msigdbr_collections</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 25 × 4
   gs_collection gs_subcollection  gs_collection_name               num_genesets
   &lt;chr&gt;         &lt;chr&gt;             &lt;chr&gt;                                   &lt;int&gt;
 1 C1            ""                Positional                                302
 2 C2            "CGP"             Chemical and Genetic Perturbati…         3538
 3 C2            "CP"              Canonical Pathways                         19
 4 C2            "CP:BIOCARTA"     BioCarta Pathways                         292
 5 C2            "CP:KEGG_LEGACY"  KEGG Legacy Pathways                      186
 6 C2            "CP:KEGG_MEDICUS" KEGG Medicus Pathways                     658
 7 C2            "CP:PID"          PID Pathways                              196
 8 C2            "CP:REACTOME"     Reactome Pathways                        1787
 9 C2            "CP:WIKIPATHWAYS" WikiPathways                              885
10 C3            "MIR:MIRDB"       miRDB                                    2377
# ℹ 15 more rows</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GO_BP</span> <span class="op">&lt;-</span> <span class="fu">msigdbr</span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Mus musculus"</span>, collection <span class="op">=</span> <span class="st">"C5"</span>, subcollection <span class="op">=</span> <span class="st">"BP"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span><span class="va">gs_name</span>, <span class="va">gene_symbol</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Using human MSigDB with ortholog mapping to mouse. Use `db_species = "MM"` for mouse-native gene sets.
This message is displayed once per session.</code></pre>
</div>
<div id="challenge" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>How many genes belong to the
<code>GOBP_REGULATION_OF_LYMPHOCYTE_MIGRATION</code> geneset?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GO_BP</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">gs_name</span> <span class="op">==</span> <span class="st">"GOBP_REGULATION_OF_LYMPHOCYTE_MIGRATION"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 69 × 2
   gs_name                                 gene_symbol
   &lt;chr&gt;                                   &lt;chr&gt;
 1 GOBP_REGULATION_OF_LYMPHOCYTE_MIGRATION Abl1
 2 GOBP_REGULATION_OF_LYMPHOCYTE_MIGRATION Abl2
 3 GOBP_REGULATION_OF_LYMPHOCYTE_MIGRATION Adam10
 4 GOBP_REGULATION_OF_LYMPHOCYTE_MIGRATION Adam17
 5 GOBP_REGULATION_OF_LYMPHOCYTE_MIGRATION Adam8
 6 GOBP_REGULATION_OF_LYMPHOCYTE_MIGRATION Adtrp
 7 GOBP_REGULATION_OF_LYMPHOCYTE_MIGRATION Aif1
 8 GOBP_REGULATION_OF_LYMPHOCYTE_MIGRATION Aire
 9 GOBP_REGULATION_OF_LYMPHOCYTE_MIGRATION Akt1
10 GOBP_REGULATION_OF_LYMPHOCYTE_MIGRATION Apod
# ℹ 59 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>Retrieve <code>Hallmarks</code> genesets</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hallmarks</span> <span class="op">&lt;-</span> <span class="fu">msigdbr</span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Mus musculus"</span>, collection <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span><span class="va">gs_name</span>, <span class="va">gene_symbol</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a></h2>
<hr class="half-width"><p>We will use the differential expression analysis comparing infected
mice at day8 and uninfected mice (day0). The following code performs
DESeq2 analysis which you should have already learnt in the previous
episode. We will focus on the genes that have an adjusted p-value (those
that have been tested).</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSet</span><span class="op">(</span><span class="va">se</span>, design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">time</span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, name <span class="op">=</span><span class="st">"time_Day8_vs_Day0"</span><span class="op">)</span></span>
<span><span class="va">res_tbl</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">res</span>, rownames <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res_tbl</span> <span class="op">&lt;-</span> <span class="va">res_tbl</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">padj</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="ora">ORA<a class="anchor" aria-label="anchor" href="#ora"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="principle">Principle<a class="anchor" aria-label="anchor" href="#principle"></a></h3>
<p>The idea behind the ORA is to test whether your list of
differentially expressed genes (DE) is enriched in certain biological
categories (such as genesets from Gene Ontology terms, KEGG pathways, or
Reactome pathways), i.e, if it contains more genes from these categories
than would be expected by chance.</p>
<div class="section level4">
<h4 id="step-1-selection-of-de-genes">Step 1: selection of DE genes<a class="anchor" aria-label="anchor" href="#step-1-selection-of-de-genes"></a></h4>
<p>To perform an over representation analysis, we first need to define
the genes we will consider as differentially expressed (DE).</p>
<figure><img src="../fig/12-GSEA-rendered-unnamed-chunk-7-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We could consider all genes that have a p-adjusted value &lt; 0.05
(left volcano), in this case we would select 7388 genes. This number is
probably too large, so we should probably be more restrictive and select
a smaller set of genes, using a more stringent filtering on the
p-adjusted value. We could also consider filtering using the logFC,
which could also make biological sense.</p>
<p>Whatever thresholds are used, the idea here is to start from a
selection of genes that we will consider as our DE genes. The selection
is a choice the user has to make.</p>
<p>Here, we will apply the selection criteria illustrated in the third
volcano. Hence, among all the genes (called the
<strong>universe</strong>), we will consider the 424 genes with a
p-adjusted value &lt; 10^{-10} and an absolute logFoldChange &gt; 0.5 as
our <strong>genes of interest</strong> or <strong>DE genes</strong>.</p>
</div>
<div class="section level4">
<h4 id="step2-counting">Step2: Counting<a class="anchor" aria-label="anchor" href="#step2-counting"></a></h4>
<p>For each biological category, we can count how many genes from our DE
genes are in that category, and how many are not.</p>
<p>Let’s say we want to test the enrichment of the
<code>GOBP_LEUKOCYTE_CHEMOTAXIS</code> geneset.</p>
<figure><img src="../fig/12-GSEA-rendered-unnamed-chunk-8-1.png" width="40%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We will count the number of DE genes from that are in
<code>GOBP_LEUKOCYTE_CHEMOTAXIS</code> geneset, and how many are
not.</p>
<table class="table"><thead><tr class="header"><th align="left"></th>
<th align="right">GO</th>
<th align="right">not_GO</th>
</tr></thead><tbody><tr class="odd"><td align="left">DE</td>
<td align="right">13</td>
<td align="right">411</td>
</tr><tr class="even"><td align="left">not_DE</td>
<td align="right">180</td>
<td align="right">23306</td>
</tr></tbody></table></div>
<div class="section level4">
<h4 id="step3-statistical-test">Step3: Statistical test<a class="anchor" aria-label="anchor" href="#step3-statistical-test"></a></h4>
<p>We can now apply a Fisher’s exact (or hypergeometric test) that will
test whether there is a significant enrichment of DE genes in the GO
term.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">fisher.test</span><span class="op">(</span><span class="va">count_mat</span>, alternative <span class="op">=</span> <span class="st">"greater"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
	Fisher's Exact Test for Count Data

data:  count_mat
p-value = 4.378e-05
alternative hypothesis: true odds ratio is greater than 1
95 percent confidence interval:
 2.362979      Inf
sample estimates:
odds ratio
  4.094913 </code></pre>
</div>
</div>
</div>
<div class="section level3">
<h3 id="running-an-ora-with-clusterprofiler-package">Running an ORA with <code>clusterProfiler</code> package<a class="anchor" aria-label="anchor" href="#running-an-ora-with-clusterprofiler-package"></a></h3>
<p>In practice, the analysis presented above can be executed using any
of the very many packages that are available.</p>
<p>Here, we will use the <a href="https://yulab-smu.top/clusterProfiler-book/" class="external-link">clusterProfiler
package</a> to test if any GO terms from the Biological Process is
enriched among our DE genes.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define our DE genes</span></span>
<span><span class="va">padj_thr</span> <span class="op">&lt;-</span> <span class="fl">1e-10</span></span>
<span><span class="va">log2FC_thr</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span><span class="va">geneList</span> <span class="op">&lt;-</span> <span class="va">res_tbl</span><span class="op">$</span><span class="va">gene</span></span>
<span><span class="va">genes_DE</span> <span class="op">&lt;-</span> <span class="va">res_tbl</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="va">res_tbl</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;=</span> <span class="va">padj_thr</span> <span class="op">&amp;</span></span>
<span>                           <span class="fu">abs</span><span class="op">(</span><span class="va">res_tbl</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">log2FC_thr</span><span class="op">]</span></span>
<span></span>
<span><span class="va">ORA_res</span> <span class="op">&lt;-</span> <span class="fu">enricher</span><span class="op">(</span>gene          <span class="op">=</span> <span class="va">genes_DE</span>,</span>
<span>                    universe      <span class="op">=</span> <span class="va">geneList</span>,</span>
<span>                    pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>                    qvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                    pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                    TERM2GENE     <span class="op">=</span> <span class="va">GO_BP</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ORA_res_tbl</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">ORA_res</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">ORA_res_tbl</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">ID</span>, <span class="va">GeneRatio</span>, <span class="va">BgRatio</span>, <span class="va">pvalue</span>, <span class="va">p.adjust</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 5
  ID                                       GeneRatio BgRatio     pvalue p.adjust
  &lt;chr&gt;                                    &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;    &lt;dbl&gt;
1 GOBP_ENSHEATHMENT_OF_NEURONS             17/284    148/13271  1.76e-8  4.66e-5
2 GOBP_MYELIN_ASSEMBLY                     7/284     20/13271   1.17e-7  1.54e-4
3 GOBP_SECONDARY_ALCOHOL_METABOLIC_PROCESS 13/284    136/13271  6.92e-6  6.12e-3
4 GOBP_ALCOHOL_METABOLIC_PROCESS           19/284    294/13271  1.89e-5  1.25e-2
5 GOBP_STEROL_BIOSYNTHETIC_PROCESS         8/284     61/13271   4.39e-5  2.19e-2
6 GOBP_STEROID_METABOLIC_PROCESS           17/284    262/13271  4.96e-5  2.19e-2</code></pre>
</div>
<p>In the output data frame, there are the following columns:</p>
<ul><li>
<code>ID</code>: ID of the gene set. In this example analysis, it is
the GO ID.</li>
<li>
<code>Description</code>: Readable description. Here it is the name
of the GO term.</li>
<li>
<code>GeneRatio</code>: Number of DE genes in the gene set / total
number of DE genes.</li>
<li>
<code>BgRatio</code>: Size of the gene set / total number of
genes.</li>
<li>
<code>pvalue</code>: <em>p</em>-value calculated from the
hypergeometric distribution.</li>
<li>
<code>p.adjust</code>: Adjusted <em>p</em>-value by the BH
method.</li>
<li>
<code>qvalue</code>: <em>q</em>-value which is another way for
controlling false positives in multiple testings.</li>
<li>
<code>geneID</code>: A list of DE genes in the gene set.</li>
<li>
<code>Count</code>: Number of DE genes in the gene set.</li>
</ul><p>You may have noticed the total number of DE genes changes. We defined
424 DE genes, but only 284 DE genes are included in the enrichment
result table (in the GeneRatio column). Similarly, our universe had
23910 genes, but the total number of tested genes was only 13271 in the
result table. The main reason is by default DE genes not annotated to
any GO gene set are filtered out.</p>
</div>
<div class="section level3">
<h3 id="visualisation">Visualisation<a class="anchor" aria-label="anchor" href="#visualisation"></a></h3>
<p>The clusterProfiler documentation provides a chapter on the <a href="https://yulab-smu.top/clusterProfiler-book/chapter12.html" class="external-link">visualization
of functional enrichment results</a> .</p>
<p>Another useful visualisation, that links the enrichment results back
to the whole set of results is to highlight the genes in a particular
set of interest on the volcano plot.</p>
<p>Let’s inspect genes from the <code>GOBP_MYELINATION</code>on the
volcano</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_geneset</span> <span class="op">&lt;-</span> <span class="st">"GOBP_ENSHEATHMENT_OF_NEURONS"</span></span>
<span><span class="va">genes_from_geneset</span> <span class="op">&lt;-</span> <span class="va">GO_BP</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">gs_name</span> <span class="op">==</span> <span class="va">my_geneset</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pull</span><span class="op">(</span><span class="va">gene_symbol</span><span class="op">)</span></span>
<span></span>
<span><span class="va">padj_thr</span> <span class="op">&lt;-</span> <span class="fl">1e-10</span></span>
<span><span class="va">log2FC_thr</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span></span>
<span><span class="va">res_tbl</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2FoldChange</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">padj</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>color <span class="op">=</span> <span class="st">"gray80"</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>data <span class="op">=</span> <span class="va">res_tbl</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">gene</span> <span class="op">%in%</span> <span class="va">genes_from_geneset</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"purple"</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">padj_thr</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">log2FC_thr</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="op">-</span><span class="va">log2FC_thr</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="fu">length</span><span class="op">(</span><span class="va">genes_DE2</span><span class="op">)</span>, <span class="st">" DE selected (padj &lt; "</span>, <span class="va">padj_thr</span>, <span class="st">")"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="va">my_geneset</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>color <span class="op">=</span> <span class="st">"purple"</span><span class="op">)</span>,</span>
<span>        axis.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/12-GSEA-rendered-unnamed-chunk-12-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
<div class="section level3">
<h3 id="conclusion-about-ora-method">Conclusion about ORA method<a class="anchor" aria-label="anchor" href="#conclusion-about-ora-method"></a></h3>
<p>This approach is straightfoward and very fast. Its major drawback
however is that we need to define a cutoff to differentiate DE from
non-DE genes. Setting this threshold might have an effect on the
results.</p>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>Test a different threshold to define the group of ‘DE genes’ and
check if, in the cases above, this has and effect on the GO terms of
interest.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define our DE genes</span></span>
<span><span class="va">padj_thr</span> <span class="op">&lt;-</span> <span class="fl">1e-20</span></span>
<span><span class="va">log2FC_thr</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span></span>
<span><span class="va">geneList</span> <span class="op">&lt;-</span> <span class="va">res_tbl</span><span class="op">$</span><span class="va">gene</span></span>
<span><span class="va">genes_DE</span> <span class="op">&lt;-</span> <span class="va">res_tbl</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="va">res_tbl</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;=</span> <span class="va">padj_thr</span> <span class="op">&amp;</span></span>
<span>                           <span class="fu">abs</span><span class="op">(</span><span class="va">res_tbl</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">log2FC_thr</span><span class="op">]</span></span>
<span></span>
<span><span class="va">res_tbl</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2FoldChange</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">padj</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">gene</span> <span class="op">%in%</span> <span class="va">genes_DE</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gray"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">padj_thr</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">log2FC_thr</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="fu">length</span><span class="op">(</span><span class="va">genes_DE</span><span class="op">)</span>, <span class="st">" DE selected (padj &lt; "</span>, <span class="va">padj_thr</span>, <span class="st">")"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/12-GSEA-rendered-unnamed-chunk-13-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ORA_res_new_thr</span> <span class="op">&lt;-</span> <span class="fu">enricher</span><span class="op">(</span>gene          <span class="op">=</span> <span class="va">genes_DE</span>,</span>
<span>                 universe      <span class="op">=</span> <span class="va">geneList</span>,</span>
<span>                 pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>                 qvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                 pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                 TERM2GENE     <span class="op">=</span> <span class="va">GO_BP</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ORA_res_new_thr_tbl</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">ORA_res_new_thr</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">ORA_res_new_thr_tbl</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">ID</span>, <span class="va">GeneRatio</span>, <span class="va">BgRatio</span>, <span class="va">pvalue</span>, <span class="va">p.adjust</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 8 × 5
  ID                                          GeneRatio BgRatio  pvalue p.adjust
  &lt;chr&gt;                                       &lt;chr&gt;     &lt;chr&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 GOBP_ENSHEATHMENT_OF_NEURONS                9/102     148/13… 1.98e-6  0.00294
2 GOBP_NEURON_PROJECTION_REGENERATION         6/102     55/132… 3.80e-6  0.00294
3 GOBP_RESPONSE_TO_AXON_INJURY                6/102     76/132… 2.52e-5  0.0130
4 GOBP_REGENERATION                           8/102     162/13… 3.44e-5  0.0133
5 GOBP_ALCOHOL_METABOLIC_PROCESS              10/102    294/13… 8.62e-5  0.0267
6 GOBP_GLUTAMINE_FAMILY_AMINO_ACID_BIOSYNTHE… 3/102     14/132… 1.51e-4  0.0389
7 GOBP_ORGANIC_ACID_BIOSYNTHETIC_PROCESS      9/102     273/13… 2.49e-4  0.0492
8 GOBP_CELLULAR_COMPONENT_ASSEMBLY_INVOLVED_… 6/102     115/13… 2.54e-4  0.0492 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">Vennerable</span><span class="op">)</span></span>
<span><span class="va">geneset_list</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span>prev_thr <span class="op">=</span> <span class="va">ORA_res_tbl</span><span class="op">$</span><span class="va">ID</span>,</span>
<span>                     new_thr <span class="op">=</span> <span class="va">ORA_res_new_thr_tbl</span><span class="op">$</span><span class="va">ID</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">Venn</span><span class="op">(</span><span class="va">geneset_list</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/12-GSEA-rendered-unnamed-chunk-13-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
</div>
</div>
</div>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<p>Repeat the ORA analysis using this time the Hallmarks genesets.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hallmarks</span> <span class="op">&lt;-</span> <span class="fu">msigdbr</span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Mus musculus"</span>, collection <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu">select</span><span class="op">(</span><span class="va">gs_name</span>, <span class="va">gene_symbol</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ORA_hallmarks</span> <span class="op">&lt;-</span> <span class="fu">enricher</span><span class="op">(</span>gene          <span class="op">=</span> <span class="va">genes_DE</span>,</span>
<span>                          universe      <span class="op">=</span> <span class="va">geneList</span>,</span>
<span>                          pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>                          qvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                          pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                          TERM2GENE     <span class="op">=</span> <span class="va">hallmarks</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ORA_hallmarks_tbl</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">ORA_hallmarks</span><span class="op">)</span> <span class="co">#%&gt;% filter(p.adjust &lt; 0.05)</span></span>
<span><span class="va">ORA_hallmarks_tbl</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">ID</span>, <span class="va">GeneRatio</span>, <span class="va">BgRatio</span>, <span class="va">pvalue</span>, <span class="va">p.adjust</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 31 × 5
   ID                               GeneRatio BgRatio  pvalue p.adjust
   &lt;chr&gt;                            &lt;chr&gt;     &lt;chr&gt;     &lt;dbl&gt;    &lt;dbl&gt;
 1 HALLMARK_XENOBIOTIC_METABOLISM   4/35      185/4049 0.0733    0.830
 2 HALLMARK_FATTY_ACID_METABOLISM   3/35      148/4049 0.134     0.830
 3 HALLMARK_PEROXISOME              2/35      95/4049  0.198     0.830
 4 HALLMARK_ANDROGEN_RESPONSE       2/35      99/4049  0.210     0.830
 5 HALLMARK_BILE_ACID_METABOLISM    2/35      101/4049 0.217     0.830
 6 HALLMARK_IL2_STAT5_SIGNALING     3/35      187/4049 0.218     0.830
 7 HALLMARK_TNFA_SIGNALING_VIA_NFKB 3/35      187/4049 0.218     0.830
 8 HALLMARK_MTORC1_SIGNALING        3/35      196/4049 0.239     0.830
 9 HALLMARK_APICAL_SURFACE          1/35      42/4049  0.307     0.830
10 HALLMARK_KRAS_SIGNALING_DN       2/35      153/4049 0.384     0.830
# ℹ 21 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="gsea">GSEA<a class="anchor" aria-label="anchor" href="#gsea"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="principle-1">Principle<a class="anchor" aria-label="anchor" href="#principle-1"></a></h3>
<p>Gene set enrichment analysis refers to a broad family of tests. Here,
we will define the principles based on <a href="https://pubmed.ncbi.nlm.nih.gov/16199517/" class="external-link">Subramanian et
al. 2005</a> keeping in mind that the exact implementation will differ
in different tools.</p>
<p>Gene Set Enrichment Analysis (GSEA) is a statistical method used to
determine whether predefined sets of genes (for example, genes belonging
to a GO term) show systematic differences in expression between two
biological conditions. Instead of focusing only on individual genes,
GSEA evaluates the <strong>collective behavior of gene
groups</strong>.</p>
<p>The major advantage of GSEA approaches is that they don’t rely on
defining DE genes.</p>
<div class="section level4">
<h4 id="step-1-gene-ranking">Step 1: Gene ranking<a class="anchor" aria-label="anchor" href="#step-1-gene-ranking"></a></h4>
<p>The first step is to order the genes of interest. Genes could be
ranked for example by the value of the test statistic or by their
p-value.</p>
<p>Depending on the metric used for the ranking, genes will be ordered
from the most up-regulated to the most down-regulated (if the test
statistic is used for instance) or by the most significantly DE (no
matter if they are up or down-regulated) to genes not differentially
expressed.</p>
</div>
<div class="section level4">
<h4 id="step-2-check-where-the-genes-of-a-specific-geneset-appear">Step 2: Check where the genes of a specific geneset appear<a class="anchor" aria-label="anchor" href="#step-2-check-where-the-genes-of-a-specific-geneset-appear"></a></h4>
<p>For a given gene set (a GO-term for example), The idea is to check if
these genes tend to appear toward the top or bottom of the ranked list,
rather than being randomly scattered.</p>
</div>
<div class="section level4">
<h4 id="step-3-compute-an-enrichment-score-es">Step 3: Compute an Enrichment Score (ES)<a class="anchor" aria-label="anchor" href="#step-3-compute-an-enrichment-score-es"></a></h4>
<p>The algorithm walks down the ranked list, increasing a running-sum
statistic when it encounters a gene from the set, and decreasing it
otherwise.</p>
<p>The positive score is defined by <span class="math inline">\(\frac{n_{genes} -
n_{genes~in~set}}{n_{genes~in~set}}\)</span>, and the decreasing score
by -1, so that the sum of all genes in the set and those not in the set
becomes zero.</p>
<p>The maximum deviation from zero is the Enrichment Score (ES),
reflecting how strongly the gene set is enriched at either end of the
ranking.</p>
<figure><img src="../fig/12-GSEA-rendered-unnamed-chunk-15-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
<div class="section level4">
<h4 id="step-4-assess-significance-by-permutation">Step 4: Assess significance by permutation<a class="anchor" aria-label="anchor" href="#step-4-assess-significance-by-permutation"></a></h4>
<p>To evaluate whether the observed ES is greater than expected by
chance, GSEA performs permutations (of either gene labels or sample
labels). This generates a null distribution to compute a p-value.</p>
<p>An Enrichment score is recomputed for each permutation. The p-value
is given by the proportion of permutations where the permuted ES is at
least as extreme as the observed ES:</p>
<p><span class="math display">\[p = \frac{\#\{ ES_{\text{permuted}} \ge
ES_{\text{observed}} \}}{\text{total number of
permutations}}\]</span></p>
</div>
</div>
<div class="section level3">
<h3 id="running-a-gsea-with-clusterprofiler-package">Running a GSEA with <code>clusterProfiler</code> package<a class="anchor" aria-label="anchor" href="#running-a-gsea-with-clusterprofiler-package"></a></h3>
<p>Here, we will re-use the <a href="https://yulab-smu.top/clusterProfiler-book/" class="external-link">clusterProfiler
package</a> to run a GSEA this time, using GO terms from the Biological
Process.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Ranking our DE results by the statistic value used in DESeq2</span></span>
<span><span class="va">geneList</span> <span class="op">&lt;-</span> <span class="va">res_tbl</span><span class="op">$</span><span class="va">stat</span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">geneList</span><span class="op">)</span><span class="op">&lt;-</span> <span class="va">res_tbl</span><span class="op">$</span><span class="va">gene</span></span>
<span><span class="va">geneList</span> <span class="op">&lt;-</span> <span class="va">geneList</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">geneList</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">GSEA_res</span> <span class="op">&lt;-</span> <span class="fu">GSEA</span><span class="op">(</span><span class="va">geneList</span>,</span>
<span>             seed <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             TERM2GENE <span class="op">=</span> <span class="va">GO_BP</span>,</span>
<span>             verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>             by <span class="op">=</span> <span class="st">"fgsea"</span>,</span>
<span>             pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>             pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>             eps <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">GSEA_res_tbl</span> <span class="op">&lt;-</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="va">GSEA_res</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">GSEA_res_tbl</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">ID</span>, <span class="va">setSize</span>, <span class="va">enrichmentScore</span>, <span class="va">NES</span>, <span class="va">pvalue</span>, <span class="va">p.adjust</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 92 × 6
   ID                             setSize enrichmentScore   NES  pvalue p.adjust
   &lt;chr&gt;                            &lt;int&gt;           &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;
 1 GOBP_CYTOPLASMIC_TRANSLATION       164           0.487  2.12 1.16e-9  3.87e-6
 2 GOBP_PROTEIN_FOLDING               200          -0.470 -2.05 1.55e-9  3.87e-6
 3 GOBP_AXON_DEVELOPMENT              499          -0.363 -1.73 1.99e-8  3.31e-5
 4 GOBP_SUBSTANTIA_NIGRA_DEVELOP…      43          -0.694 -2.34 2.81e-8  3.51e-5
 5 GOBP_ENSHEATHMENT_OF_NEURONS       148          -0.490 -2.05 4.72e-8  4.71e-5
 6 GOBP_NEURAL_NUCLEUS_DEVELOPME…      63          -0.614 -2.24 2.41e-7  2.01e-4
 7 GOBP_MIDBRAIN_DEVELOPMENT           87          -0.545 -2.10 6.88e-7  4.91e-4
 8 GOBP_REGULATION_OF_ADAPTIVE_I…     183           0.426  1.89 9.67e-7  5.62e-4
 9 GOBP_REGULATION_OF_LEUKOCYTE_…     205           0.409  1.84 1.01e-6  5.62e-4
10 GOBP_REGULATION_OF_LYMPHOCYTE…     149           0.454  1.95 1.37e-6  6.87e-4
# ℹ 82 more rows</code></pre>
</div>
<p>In the output data frame, there are the following columns:</p>
<ul><li>
<code>ID</code>: ID of the gene set. In this example analysis, it is
the GO ID.</li>
<li>
<code>Description</code>: Readable description. Here it is the name
of the GO term.</li>
<li>
<code>setSize</code>: Number of genes in this gene set that are
present in your universe of tested genes.</li>
<li>
<code>enrichmentScore</code>: Raw enrichment score calculated from
the running-sum statistic.</li>
<li>
<code>NES</code>: ES normalized by gene set size. T</li>
<li>
<code>pvalue</code>: <em>p</em>-value calculated from the
hypergeometric distribution.</li>
<li>
<code>p.adjust</code>: Adjusted <em>p</em>-value by the BH
method.</li>
<li>
<code>qvalue</code>: <em>q</em>-value which is another way for
controlling false positives in multiple testings.</li>
<li>
<code>rank</code>: Position in the ranked gene list where the
running score reaches its maximum (the ES).</li>
<li>
<code>leadingEdge</code>: List of genes in the set contributing most
to the ES (genes before the peak), often called the core
enrichment.</li>
</ul><p>The GSEA plot can be drawn with the gseaplot() function</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">gseaplot</span><span class="op">(</span><span class="va">GSEA_res</span>, <span class="va">GSEA_res_tbl</span><span class="op">$</span><span class="va">ID</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, by <span class="op">=</span> <span class="st">"runningScore"</span>, title <span class="op">=</span> <span class="va">GSEA_res_tbl</span><span class="op">$</span><span class="va">ID</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/12-GSEA-rendered-unnamed-chunk-17-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="challenge-4" class="callout-inner">
<h3 class="callout-title">Challenge:</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li><p>Think about the different ways to order the genes for the GSEA
analysis. What impact could the ranking method have on the GSEA
analysis?</p></li>
<li><p>Compare GSEA and ORA methods, what are pros and cons of both
methods?</p></li>
<li><p>Compare the results obtained using ORA and GSEA on GO terms
(BP)</p></li>
</ol></div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">Vennerable</span><span class="op">)</span></span>
<span><span class="va">geneset_list</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span>ORA <span class="op">=</span> <span class="va">ORA_res_tbl</span><span class="op">$</span><span class="va">ID</span>,</span>
<span>                     GSEA <span class="op">=</span> <span class="va">GSEA_res_tbl</span><span class="op">$</span><span class="va">ID</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">Venn</span><span class="op">(</span><span class="va">geneset_list</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/12-GSEA-rendered-unnamed-chunk-18-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
</div>
</div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/11-Additional-material-PCA.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/11-Additional-material-PCA.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Additional material:</a>
        <a class="chapter-link float-end" href="../instructor/index.html" rel="next">Next... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/12-GSEA.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/bioc-rnaseq/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/bioc-rnaseq/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/bioc-rnaseq/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:charlotte.soneson@fmi.ch">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="../LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>,
        <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>,
      and <a href="https://github.com/mblue9/varnish/tree/3403ad0dd37f7641c6442330fcc182937e7964c5" class="external-link">varnish (0.2.16)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://UCLouvain-BIOINFO.github.io/bioc-rnaseq/instructor/12-GSEA.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Additional material: GSEA",
  "creativeWorkStatus": "active",
  "url": "https://UCLouvain-BIOINFO.github.io/bioc-rnaseq/instructor/12-GSEA.html",
  "identifier": "https://UCLouvain-BIOINFO.github.io/bioc-rnaseq/instructor/12-GSEA.html",
  "dateCreated": "2020-09-15",
  "dateModified": "2025-11-16",
  "datePublished": "2025-11-19"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

